File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        subProblem = [float('inf')] * (amount + 1)
        subProblem[0] = 0

        for coin in coins:
            for num in range(coin, amount+1):
                subProblem[num] = min(subProblem[num], subProblem[num - coin] + 1)

        return -1 if subProblem[amount] == float('inf') else subProblem[amount]

File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int []dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        dp[0] = 0;
        for(int idx = 1; idx < amount+1;idx++){
            for(int coin : coins){
                if(idx - coin >= 0){
                    dp[idx] = Math.min(dp[idx], 1 + dp[idx - coin]);
                }
            }
        }        
        if(dp[amount] != amount + 1){
            return dp[amount];
        }
        return -1;    
    }
}
